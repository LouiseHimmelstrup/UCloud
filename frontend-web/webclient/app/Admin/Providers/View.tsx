import * as React from "react";
import {InvokeCommand, useCloudCommand} from "@/Authentication/DataHook";
import * as UCloud from "@/UCloud";
import {useCallback, useState} from "react";
import {Box, Button, Label, Select} from "@/ui-components";
import {inDevEnvironment, onDevSite, PropType} from "@/UtilityFunctions";
import {Operation} from "@/ui-components/Operation";
import {addStandardDialog} from "@/UtilityComponents";
import {auth, BulkResponse} from "@/UCloud";
import {bulkRequestOf} from "@/DefaultObjects";
import AccessToken = auth.AccessToken;
import ResourceForm from "@/Products/CreateProduct";
import * as Types from "@/Accounting";
import {Provider} from "@/UCloud/ProvidersApi";

const productTypes: {value: PropType<Types.Product, "type">, title: string}[] = [
    {value: "compute", title: "Compute"},
    {value: "network_ip", title: "Public IP"},
    {value: "ingress", title: "Public link"},
    {value: "license", title: "Software license"},
    {value: "storage", title: "Storage"},
];

function unitNameFromType(type: PropType<Types.Product, "type">): string {
    let unitName = "";
    switch (type) {
        case "compute":
            unitName = "minute";
            break;
        case "storage":
            unitName = "GB/day";
            break;
        case "network_ip":
        case "license":
        case "ingress":
            unitName = "activation";
            break;
    }
    return unitName;
}

export const ProductCreationForm: React.FunctionComponent<{provider: Provider, onComplete: () => void}> = props => {
    const [type, setType] = useState<PropType<Types.Product, "type">>("compute");
    const onTypeChange = useCallback(e => setType(e.target.value as PropType<Types.Product, "type">), [setType]);
    const [licenseTagCount, setTagCount] = useState(1);
    const [, invokeCommand] = useCloudCommand();

    const unitName = unitNameFromType(type);

    return <Box maxWidth={"800px"} margin={"0 auto"}>
        <Label>
            Type
            <Select value={type} onChange={onTypeChange}>
                {productTypes.map(it => <option key={it.value} value={it.value}>{it.title}</option>)}
            </Select>
        </Label>

        <ResourceForm
            title="Product"
            createRequest={async (data): Promise<APICallParameters<any>> => {
                const tokens = await invokeCommand<BulkResponse<AccessToken>>(UCloud.auth.providers.refresh(
                    bulkRequestOf({refreshToken: props.provider.refreshToken}))
                );

                const accessToken = tokens?.responses[0]?.accessToken;
                let product: Types.Product;

                const shared: Omit<Types.ProductBase, "productType"> = {
                    type,
                    category: {name: data.fields.name, provider: props.provider.id},
                    pricePerUnit: data.fields.pricePerUnit * 10_000,
                    name: data.fields.name,
                    description: data.fields.description,
                    priority: data.fields.priority,
                    version: data.fields.version,
                    freeToUse: data.fields.freeToUse,
                    unitOfPrice: data.fields.unitOfPrice,
                    chargeType: data.fields.chargeType,
                    hiddenInGrantApplications: data.fields.hiddenInGrantApplications,
                };

                const tags: string[] = [];
                for (let i = 0; i < licenseTagCount; i++) {
                    const entry = data.fields[`tag-${i}`];
                    if (entry) tags.push(entry);
                }

                switch (type) {
                    case "storage":
                        product = {
                            ...shared,
                            productType: "STORAGE"
                        } as Types.ProductStorage;
                        break;
                    case "compute":
                        product = {
                            ...shared,
                            productType: "COMPUTE",
                            cpu: data.fields.cpu,
                            memoryInGigs: data.fields.memory,
                            gpu: data.fields.gpu
                        } as Types.ProductCompute;
                        break;
                    case "ingress":
                        product = {
                            ...shared,
                            productType: "INGRESS",
                        } as Types.ProductIngress;
                        break;
                    case "network_ip":
                        product = {
                            ...shared,
                            productType: "NETWORK_IP"
                        } as Types.ProductNetworkIP;
                        break;
                    case "license":
                        product = {
                            ...shared,
                            productType: "LICENSE",
                            tags
                        } as Types.ProductLicense;
                        break;
                }

                return {...UCloud.accounting.products.createProduct(bulkRequestOf(product as any)), accessTokenOverride: accessToken};
            }}
        >
            <ResourceForm.Text required id="name" placeholder="Name..." label="Name (e.g. u1-standard-1)" styling={{}} />
            <ResourceForm.Number required id="pricePerUnit" placeholder="Price..." rightLabel="DKK" label={`Price per ${unitName}`} step="0.01" min={0} styling={{}} />
            <ResourceForm.TextArea required id="description" placeholder="Description..." label="Description" rows={10} styling={{}} />
            <ResourceForm.Number required id="priority" placeholder="Priority..." label="Priority" styling={{}} />
            <ResourceForm.Number required id="version" placeholder="Version..." label="Version" min={0} styling={{}} />
            <ResourceForm.Checkbox id="freeToUse" defaultChecked={false} label="Free to use" styling={{}} />
            <ResourceForm.Select id="unitOfPrice" label="Unit of Price" required options={[
                {value: "PER_UNIT", text: "Per Unit"},
                {value: "CREDITS_PER_MINUTE", text: "Credits Per Minute"},
                {value: "CREDITS_PER_HOUR", text: "Credits Per Hour"},
                {value: "CREDITS_PER_DAY", text: "Credits Per Day"},
                {value: "UNITS_PER_MINUTE", text: "Units Per Minute"},
                {value: "UNITS_PER_HOUR", text: "Units Per Hour"},
                {value: "UNITS_PER_DAY", text: "Units Per Day"}
            ]} styling={{}} />
            <ResourceForm.Select id="chargeType" label="Chargetype" required options={[
                {value: "ABSOLUTE", text: "Absolute"},
                {value: "DIFFERENTIAL_QUOTA", text: "Differential Quota"}
            ]} styling={{}} />
            <ResourceForm.Checkbox id="hiddenInGrantApplications" label="Hidden in Grant Applications" defaultChecked={false} styling={{}} />

            {type !== "compute" ? null : (
                <>
                    <ResourceForm.Number id="cpu" placeholder="vCPU..." label="vCPU" required styling={{}} />
                    <ResourceForm.Number id="memory" placeholder="Memory..." label="Memory in GB" required styling={{}} />
                    <ResourceForm.Number id="gpus" placeholder="GPUs..." label="Number of GPUs" required styling={{}} />
                </>
            )}
            {type !== "license" ? null : (
                <>
                    {[...Array(licenseTagCount).keys()].map(id =>
                        <ResourceForm.Text key={id} id={`tag-${id}`} label={`Tag ${id + 1}`} styling={{}} />
                    )}
                    <div>
                        <Button fullWidth type="button" onClick={() => setTagCount(t => t + 1)} mt="6px">Add tag</Button>
                    </div>
                </>
            )}
        </ResourceForm>
    </Box>;
};

interface OpCallbacks {
    provider: string;
    invokeCommand: InvokeCommand;
    reload: () => void;
    startProductCreation: () => void;
    stopProductCreation: () => void;
    isCreatingProduct: boolean;
}

const operations: Operation<UCloud.provider.Provider, OpCallbacks>[] = [
    {
        enabled: (_, cb) => !cb.isCreatingProduct,
        onClick: (_, cb) => {
            cb.startProductCreation();
        },
        text: "Create product",
        operationType: () => Button,
    },
    {
        enabled: selected => selected.length > 0,
        onClick: (selected, cb) =>
            addStandardDialog({
                title: "WARNING!",
                message: <>
                    <p>Are you sure you want to renew the provider token?</p>
                    <p>
                        This will invalidate every current security token. Your provider <i>must</i> be reconfigured
                        to use the new tokens.
                    </p>
                </>,
                confirmText: "Confirm",
                cancelButtonColor: "blue",
                confirmButtonColor: "red",
                cancelText: "Cancel",
                onConfirm: async () => {
                    await cb.invokeCommand(UCloud.provider.providers.renewToken(
                        {type: "bulk", items: selected.map(it => ({id: it.id}))}
                    ));

                    cb.reload();
                }
            }),
        text: "Renew token",
        color: "red",
        icon: "trash",
        operationType: () => Button,
    },
    {
        enabled: selected => selected.length === 1 && (inDevEnvironment() || onDevSite()),
        onClick: async (selected, cb) => {
            await cb.invokeCommand(
                UCloud.accounting.wallets.grantProviderCredits({provider: cb.provider})
            );
            cb.reload();
        },
        text: "Grant credits",
        operationType: () => Button
    }
];
