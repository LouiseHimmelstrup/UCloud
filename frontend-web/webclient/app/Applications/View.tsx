import * as React from "react";
import PromiseKeeper from "PromiseKeeper";
import { Cloud } from "Authentication/SDUCloudObject";
import { Link } from "react-router-dom";
import * as ReactMarkdown from "react-markdown";
import { DefaultLoading } from "LoadingIcon/LoadingIcon";
import { ApplicationInformation, Application } from "Applications";
import { Image, OutlineButton, Button, Box, VerticalButtonGroup} from "ui-components";
import * as Heading from "ui-components/Heading"
import styled from "styled-components";
import { dateToString } from "Utilities/DateUtilities";
import { toLowerCaseAndCapitalize } from "UtilityFunctions"
import ContainerForText from "ui-components/ContainerForText";
import { MainContainer } from "MainContainer/MainContainer";
import { ApplicationCardContainer, SlimApplicationCard } from "./Card";
import { Page } from "Types";
import { promises } from "fs";

const circuitBoard = require("Assets/Images/circuitboard-bg.png");

type DetailedApplicationProps = any
type DetailedApplicationState = {
    appInformation?: ApplicationInformation
    previousVersions?: Page<Application>
    promises: PromiseKeeper
    loading: boolean
    error?: string
}

interface MainContentProps {
    onFavorite?: () => void
    application: ApplicationInformation
}

export class View extends React.Component<DetailedApplicationProps, DetailedApplicationState> {
    constructor(props) {
        super(props);
        this.state = {
            promises: new PromiseKeeper,
            loading: false,
            appInformation: undefined,
            error: undefined
        }
    }

    componentDidMount() {
        this.retrieveApplication();
        this.retrievePreviousVersions();
    }

    retrieveApplication() {
        this.setState(() => ({ loading: true }));
        const { appName, appVersion } = this.props.match.params;
        const { promises } = this.state;
        promises.makeCancelable(Cloud.get(`/hpc/apps/${encodeURI(appName)}/${encodeURI(appVersion)}`))
            .promise.then(({ response }: { response: ApplicationInformation }) =>
                this.setState(() => ({
                    appInformation: response,
                    loading: false,
                }))
            ).catch(_ => this.setState({
                error: `An error occurred fetching ${appName}`,
                loading: false
            }));
    }

    retrievePreviousVersions() {
        const { appName } = this.props.match.params;
        const { promises } = this.state;
        promises.makeCancelable(Cloud.get(`/hpc/apps/${encodeURI(appName)}`))
            .promise.then(({ response }: { response: Page<Application> }) =>
                this.setState(() => ({
                    previousVersions: response,
                }))
            ).catch(_ => this.setState({
                error: `An error occurred fetching versions of ${appName}`,
            }));
    }

    render() {
        const { appInformation } = this.state;
        return (
            <MainContainer
                header={
                    !appInformation ? null :
                        <AppHeader application={appInformation} />
                }

                main={!appInformation ?
                    <DefaultLoading loading={this.state.loading} /> :
                    <ContainerForText>
                        <Content
                            application={appInformation}
                            previousVersions={this.state.previousVersions} />
                    </ContainerForText>
                }

                sidebar={
                    !appInformation ? null : <Sidebar application={appInformation} />
                }
            />
        );
    }
}

const AppHeaderBase = styled.div`
    display: flex;
    flex-direction: row;

    & > ${Image} {
        width: 128px;
        height: 128px;
        border-radius: 8px;
        object-fit: cover;
        margin-right: 16px;
    }
`;

const AppHeaderDetails = styled.div`
    display: flex;
    flex-direction: column;

    & > h1, h2 {
        margin: 0;
    }
`;

const AppSection = styled(Box)`
    margin-bottom: 16px;
`;

const AppHeader: React.StatelessComponent<MainContentProps> = props => (
    <AppHeaderBase>
        <Image src={circuitBoard} />
        <AppHeaderDetails>
            <h1>{props.application.description.title}</h1>
            <h2>v{props.application.description.info.version}</h2>
            <span>{props.application.description.authors.join(", ")}</span>
            <Tags tags={["Test Tag 1", "Test Tag 2", "Test Tag 3"]} />
        </AppHeaderDetails>
    </AppHeaderBase>
);

const Sidebar: React.StatelessComponent<MainContentProps> = props => (
    <>
        <VerticalButtonGroup>
            <Button color={"blue"}>Add to My Applications</Button>
            <Link to={`/applications/${props.application.description.info.name}/${props.application.description.info.version}`}><OutlineButton color={"blue"}>Run Application</OutlineButton></Link>
            <a target="_blank" href="https://duckduckgo.com" rel="noopener"><OutlineButton color={"blue"}>Website</OutlineButton></a>
        </VerticalButtonGroup>
    </>
);

function Content(props: MainContentProps & { previousVersions?: Page<Application> }): JSX.Element {
    return (
        <>
            <AppSection>
                <ReactMarkdown
                    unwrapDisallowed
                    source={props.application.description.description}
                    disallowedTypes={[
                        "image",
                        "heading"
                    ]}
                />
            </AppSection>

            <AppSection>
                <PreviousVersions previousVersions={props.previousVersions} />
            </AppSection>

            <AppSection>
                <Information application={props.application} />
            </AppSection>
        </>
    );
}

const PreviousVersions: React.StatelessComponent<{ previousVersions?: Page<Application> }> = props => (
    <>
        <Heading.h4>Previous Versions</Heading.h4>
        {!props.previousVersions ? null :
            <ApplicationCardContainer>
                {props.previousVersions.items.map((it, idx) => (
                    <SlimApplicationCard linkToRun app={it} key={idx} />
                ))}
            </ApplicationCardContainer>
        }
    </>
);


const TagStyle = styled.a`
    text-decoration: none;
    padding: 6px;
    margin-right: 3px;
    border: 1px solid ${props => props.theme.colors.gray};
    border-radius: 5px;
`;

const TagBase = styled.div`
    display: flex;
    flex-direction: row;
`;

function Tags({ tags }: { tags: string[] }) {
    if (!tags) return null;

    return <div>
        <TagBase>
            {
                tags.slice(0, 5).map(tag => (
                    <TagStyle href={`foo/${tag}`}>{tag}</TagStyle>
                ))
            }
        </TagBase>
    </div>;
}

function InfoAttribute(props: {
    name: string,
    value?: string,
    children?: JSX.Element
}) {
    return <Box mb={8} mr={32}>
        <Heading.h5>{props.name}</Heading.h5>
        {props.value}
        {props.children}
    </Box>;
}

const pad = (value, length) =>
    (value.toString().length < length) ? pad("0" + value, length) : value;

const InfoAttributes = styled.div`
    display: flex;
    flex-direction: row;
`;

function Information({ application }: { application: ApplicationInformation }) {
    const time = application.tool.description.defaultMaxTime;
    const timeString = `${pad(time.hours, 2)}:${pad(time.minutes, 2)}:${pad(time.seconds, 2)}`;

    return <>
        <Heading.h4>Information</Heading.h4>

        <InfoAttributes>
            <InfoAttribute
                name="Release Date"
                value={dateToString(application.createdAt)} />

            <InfoAttribute
                name="Default Time Allocation"
                value={timeString} />

            <InfoAttribute
                name="Default Nodes"
                value={`${application.tool.description.defaultNumberOfNodes}`} />

            <InfoAttribute
                name="Container Type"
                value={toLowerCaseAndCapitalize(application.tool.description.backend)} />
        </InfoAttributes>
    </>;
}