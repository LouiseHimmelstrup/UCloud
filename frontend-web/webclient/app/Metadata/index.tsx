import { History } from "history";
import { Creator, Subject, RelatedIdentifier } from "./api";

export { CreateUpdate } from "./CreateUpdate";
export { Search } from "./Search";
export { ManagedView, View } from "./View";

export interface CreateUpdateState {
    id?: string
    path: string
    title: string
    description: string
    license: {}
    keywords: string[]
    notes: ""
    dataManagementPlan: ""
    contributors: Creator[]
    references: string[]
    grants: string[]
    subjects: Subject[]
    relatedIdentifiers: RelatedIdentifier[]
    errors: { 
        description?: boolean
        title?: boolean
        contributors: {}
        subjects: {}
        relatedIdentifiers: {}
    }
}

export interface CreateUpdateProps {
    match: {
        params: string[]
    }
    history: History
}