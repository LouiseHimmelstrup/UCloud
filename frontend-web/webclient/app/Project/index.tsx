import { Contributor, Subject, RelatedIdentifier } from "./api";
import { RouterLocationProps } from "Utilities/URIUtilities";
import { AddSnackOperation } from "Snackbar/Snackbars";
export { default as CreateUpdate } from "./CreateUpdate";
export { ManagedView, View } from "./View";
export { default as Manage } from "./Management"

export interface CreateUpdateState {
    id?: string
    path: string
    title: string
    description: string
    license?: { title: string, link: string, identifier: string }
    keywords: string[]
    notes: string
    dataManagementPlan: string
    contributors: Contributor[]
    references: string[]
    grants: string[]
    subjects: Subject[]
    relatedIdentifiers: RelatedIdentifier[]
    errors: {
        description?: boolean
        title?: boolean
        contributors: {}
        subjects: {}
        relatedIdentifiers: {}
    }
}

export type CreateUpdateProps = RouterLocationProps & AddSnackOperation;

export interface DetailedProjectSearchReduxState {
    error?: string
    loading: boolean
    hidden: boolean
    projectName: string
}

export interface DetailedProjectSearchOperations {
    setProjectName: (name: string) => void
    fetchProjectsFromName: (name: string, itemsPerPage: number, page: number, callback: () => void) => void
    setError: (error?: string) => void
}