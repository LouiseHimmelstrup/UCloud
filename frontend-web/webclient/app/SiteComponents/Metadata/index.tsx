export { CreateUpdate } from "./CreateUpdate";
export { Search } from "./Search";
export { ManagedView, View } from "./View";