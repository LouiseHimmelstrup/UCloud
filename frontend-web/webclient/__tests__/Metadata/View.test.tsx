import * as React from "react";
import { View } from "Project/View";
import { create } from "react-test-renderer";
import { metadata } from "../mock/Metadata";
import { MemoryRouter } from "react-router";
import { Provider } from "react-redux";
import { configureStore, responsive } from "Utilities/ReduxUtilities";
import "jest-styled-components";

describe("View component", () => {
    test("Mount view", () =>
        expect(create(
            <Provider store={configureStore({}, { responsive })}>
                <MemoryRouter>
                    <View metadata={metadata.metadata} canEdit={true} />
                </MemoryRouter>
            </Provider>
        )).toMatchSnapshot()
    );
}) 