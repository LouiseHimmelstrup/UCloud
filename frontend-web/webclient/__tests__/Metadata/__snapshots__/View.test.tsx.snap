// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View component Mount view 1`] = `
.c0 {
  cursor: auto;
  margin-left: 190px;
  padding-bottom: 14px;
  padding-left: 14px;
  padding-right: 0;
  padding-top: 48px;
}

.c5 {
  cursor: auto;
}

.c9 {
  cursor: auto;
  margin-bottom: 3em;
}

.c20 {
  cursor: auto;
  padding-right: 240px;
  padding-top: 96px;
}

.c3 {
  cursor: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  cursor: inherit;
  font-size: 32px;
  margin: 0px;
}

.c10 {
  cursor: inherit;
  font-size: 20px;
  margin: 0px;
}

.c11 {
  cursor: inherit;
}

.c15 {
  cursor: inherit;
  color: white;
}

.c19 {
  cursor: inherit;
  margin-left: 4px;
  margin-right: 6px;
}

.c8 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  vertical-align: middle;
  cursor: inherit;
  margin-right: 0.5em;
  margin-top: 0.15em;
  margin-bottom: 0.15em;
}

.c12 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  vertical-align: middle;
  cursor: inherit;
  margin-top: 0.2em;
  margin-bottom: 0.2em;
}

.c16 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  vertical-align: middle;
  cursor: inherit;
  margin-right: 0.5em;
}

.c7 {
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 700;
  line-height: 1.5;
  cursor: pointer;
  border-radius: 5px;
  background-color: #0055d5;
  color: #fff;
  border-width: 0;
  border-style: solid;
  -webkit-transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  width: 100%;
  font-size: 14px;
  padding: 9.5px 18px;
}

.c7:disabled {
  opacity: 0.25;
}

.c7:focus {
  outline: none;
}

.c7:hover {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c14 {
  -webkit-font-smoothing: antialiased;
  display: inline-block;
  vertical-align: middle;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: inherit;
  font-weight: 700;
  line-height: 0.3;
  cursor: pointer;
  border-radius: 5px;
  background-color: #0055d5;
  color: #fff;
  border-width: 0;
  border-style: solid;
  -webkit-transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  transition: cubic-bezier(0.5,0,0.25,1) 60ms;
  width: 100%;
  font-size: 10px;
  padding: 7px 12px;
}

.c14:disabled {
  opacity: 0.25;
}

.c14:focus {
  outline: none;
}

.c14:hover {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c1 {
  cursor: auto;
  padding-left: 204px;
  padding-right: 14px;
  padding-top: 14px;
  padding-bottom: 14px;
  width: 100%;
  height: 96px;
  background-color: #fff;
  position: absolute;
  top: 48px;
  left: 0;
  position: fixed;
  z-index: 50;
}

.c4 {
  cursor: auto;
  padding-top: 144px;
  padding-left: 14px;
  padding-right: 14px;
  width: 240px;
  height: 100%;
  position: absolute;
  top: 0;
  right: 0;
  overflow-y: auto;
  position: fixed;
  z-index: 45;
}

.c6 {
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: text;
}

.c6:hover {
  color: textHighlight;
}

.c18 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: top;
  min-height: 24px;
  width: 100%;
  font-weight: 600;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  border-radius: 4px;
  border-width: 1px;
  border-style: solid;
  background-color: #ebeff3;
  border-color: #ebeff3;
  color: #1e252e;
  margin-bottom: 0.4em;
  margin-right: 4px;
  padding-left: 4px;
  padding-right: 4px;
  padding-top: 0px;
  padding-bottom: 0px;
  font-size: 10px;
}

.c13 {
  padding-bottom: 1em;
  font-size: large;
}

.c17 {
  margin-bottom: 1em;
  font-size: large;
}

<div
  className="c0"
  cursor="auto"
>
  <div
    className="c1"
    cursor="auto"
    height={96}
    width={1}
  >
    <h2
      className="c2"
      cursor="inherit"
      fontSize={5}
    >
      Untitled project
    </h2>
    <div
      className="c3"
      cursor="inherit"
    />
  </div>
  <div
    className="c4"
    cursor="auto"
    height="100%"
    width={240}
  >
    <div
      className="c5"
      cursor="auto"
    >
      <a
        className="c6"
        color="text"
        hoverColor="textHighlight"
        href="/projects/edit?filePath=%2Fhome%2Fjonas%40hinchely.dk%2FThe%20Project"
        onClick={[Function]}
      >
        <button
          className="c7"
          color="blue"
        >
          <svg
            className="c8"
            clipRule="evenodd"
            cursor="inherit"
            fill="currentcolor"
            fillRule="evenodd"
            height="1em"
            mr="0.5em"
            my="0.15em"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M23.6 5.4a1.289 1.289 0 0 0 0-1.867L20.467.4A1.289 1.289 0 0 0 18.6.4l-2.467 2.467 5 5L23.6 5.4z"
              fill={null}
              fillRule="nonzero"
            />
            <path
              d="M0 19v5h5L19.733 9.267l-5-5L0 19z"
              fillRule="nonzero"
            />
          </svg>
          Edit
        </button>
      </a>
      <div
        className="c9"
        cursor="auto"
      />
      <h4
        className="c10"
        cursor="inherit"
        fontSize={3}
      >
        <span
          className="c11"
          cursor="inherit"
        >
          <svg
            className="c12"
            clipRule="evenodd"
            color2="#ebeff3"
            cursor="inherit"
            fill="currentcolor"
            fillRule="evenodd"
            height="1.1em"
            my="0.2em"
            viewBox="0 0 25 25"
            width="1.1em"
          >
            <circle
              cx={424.005}
              cy={664.005}
              fill="#ebeff3"
              r={12.005}
              transform="matrix(1 0 0 1 -412 -652)"
            />
            <text
              fontFamily="'IBMPlexSerif-SemiBold','IBM Plex Serif SemiBold',serif"
              fontSize={24}
              fontWeight={600}
              transform="matrix(1 0 0 1 -579.022 -609.246)"
              x={586.816}
              y={629.892}
            >
              i
            </text>
          </svg>
           
          About
        </span>
      </h4>
      <div
        className="c13"
        fontSize="large"
      >
        <a
          className="c6"
          color="text"
          href="https://spdx.org/licenses/AML.html"
          rel="noopener"
          target="_blank"
        >
          <button
            className="c14"
            color="blue"
            size="small"
          >
            <div
              className="c15"
              color="white"
              cursor="inherit"
            >
              <svg
                className="c16"
                clipRule="evenodd"
                cursor="inherit"
                fill="currentcolor"
                fillRule="evenodd"
                height="1.5em"
                mr="0.5em"
                viewBox="0 0 24 24"
                width="1.5em"
              >
                <path
                  d="M21 24.001H3a3 3 0 0 1-3-3v-18a3 3 0 0 1 3-3h18a3 3 0 0 1 3 3v18a3.004 3.004 0 0 1-3 3zm-3-3l3-3v-12l-3-3H6l-3 3v12l3 3h12z"
                />
                <path
                  d="M16.91 17.894H7.09c-.602 0-1.09-.648-1.09-1.447S6.488 15 7.09 15h9.82c.602 0 1.09.648 1.09 1.447s-.488 1.447-1.09 1.447z"
                  fill={null}
                  fillRule="nonzero"
                />
                <path
                  d="M11.952 5.728l1.902 1.382-.951 1.31 1.538.5-.726 2.235-1.54-.5v1.618h-2.35v-1.618l-1.54.5L7.56 8.92l1.538-.5-.951-1.309 1.902-1.382.95 1.31.953-1.31z"
                  fill={null}
                />
              </svg>
              AML
               License
            </div>
          </button>
        </a>
      </div>
      <h4
        className="c10"
        cursor="inherit"
        fontSize={3}
      >
        <span
          className="c11"
          cursor="inherit"
        >
          <svg
            className="c12"
            clipRule="evenodd"
            color2="#ebeff3"
            cursor="inherit"
            fill="currentcolor"
            fillRule="evenodd"
            height="1.1em"
            my="0.2em"
            viewBox="0 0 24 24"
            width="1.1em"
          >
            <path
              d="M21.964 3.404v8.135l-11.4 11.752c.429.461 1.114.71 1.543.71.429 0 1.221-.214 1.682-.71l10.21-10.368V5.54l-2.035-2.135z"
              fill="#ebeff3"
              fillRule="nonzero"
            />
            <path
              d="M18.857 0h-8.571L.659 10.823A2.452 2.452 0 0 0 0 12.462c-.016.652.198 1.344.66 1.84l6.631 7.13c.429.462 1.114.722 1.543.722.428 0 1.216-.225 1.677-.721l10.06-10.356v-9.23L18.857 0zm-1.645 5.925C16.05 6.277 15 5.348 15 4.155c0-1.022.766-1.847 1.714-1.847 1.109 0 1.971 1.13 1.645 2.383-.156.594-.595 1.067-1.147 1.234z"
              fillRule="nonzero"
            />
          </svg>
           
          Keywords
        </span>
      </h4>
      <div
        className="c17"
        fontSize="large"
      >
        <div
          className="c18"
          color="lightGray"
          fontSize={0}
        >
          <div
            className="c19"
            cursor="inherit"
          >
            Example
          </div>
        </div>
      </div>
      <h4
        className="c10"
        cursor="inherit"
        fontSize={3}
      >
        <span
          className="c11"
          cursor="inherit"
        >
          <svg
            className="c12"
            clipRule="evenodd"
            color2="#ebeff3"
            cursor="inherit"
            fill="currentcolor"
            fillRule="evenodd"
            height="1.1em"
            my="0.2em"
            viewBox="0 0 24 24"
            width="1.1em"
          >
            <path
              d="M4 24h3l4-24H8L4 24zM13 24h3l4-24h-3l-4 24z"
            />
            <path
              d="M21.993 18H0l.5-3h21.993l-.5 3zM23.493 9H1.5L2 6h21.993l-.5 3z"
            />
          </svg>
           
          References
        </span>
      </h4>
      <div
        className="c17"
        fontSize="large"
      />
      <h4
        className="c10"
        cursor="inherit"
        fontSize={3}
      >
        <span
          className="c11"
          cursor="inherit"
        >
          <svg
            className="c12"
            clipRule="evenodd"
            color2="#ebeff3"
            cursor="inherit"
            fill="currentcolor"
            fillRule="evenodd"
            height="1.1em"
            my="0.2em"
            viewBox="0 0 24 18"
            width="1.1em"
          >
            <path
              d="M0 0v13.714h24V0H0zm8.598 12H4.291a2.574 2.574 0 0 0-2.576-2.577v-4.28a3.432 3.432 0 0 0 3.428-3.429h3.455c-1.066 1.26-1.74 3.097-1.74 5.143s.674 3.884 1.74 5.143zm13.688-2.577A2.59 2.59 0 0 0 19.715 12h-4.313c1.066-1.26 1.741-3.097 1.741-5.143s-.675-3.884-1.74-5.143h3.454a3.432 3.432 0 0 0 3.429 3.429v4.28z"
              fillRule="nonzero"
            />
            <path
              d="M0 15.428h24v1.715H0z"
              fill="#ebeff3"
              fillRule="nonzero"
            />
          </svg>
           
          Grants
        </span>
      </h4>
      <div
        className="c17"
        fontSize="large"
      />
    </div>
  </div>
  <div
    className="c20"
    cursor="auto"
  >
    <p>
      Project description goes here...
    </p>
  </div>
</div>
`;
